tests/TestCase/ORM/Association/HasManyTest.php

```diff
@@ -740,4 +740,216 @@ public function testLinkUsesSingleTransaction()
         $new = $this->author->get(2, ['contain' => 'Articles']);
         $this->assertCount(3, $new->articles);
     }
+
+    /**
+     * Test that saveAssociated() fails on non-empty, non-iterable value
+     *
+     * @expectedException \InvalidArgumentException
+     * @expectedExceptionMessage Could not save comments, it cannot be traversed
+     * @return void
+     */
+    public function testSaveAssociatedNotEmptyNotIterable()
+    {
+        $articles = TableRegistry::get('Articles');
+        $assoc = $articles->hasMany('Comments', [
+            'saveStrategy' => HasMany::SAVE_APPEND
+        ]);
+
+        $entity = new Entity(
+            [
+                'id' => 1,
+                'comments' => 'oh noes',
+            ],
+            ['markNew' => true]
+        );
+
+        $assoc->saveAssociated($entity);
+    }
+
+    /**
+     * Provider for empty values
+     *
+     * @return array
+     */
+    public function emptyProvider()
+    {
+        return [
+            [''],
+            [false],
+            [null],
+            [[]]
+        ];
+    }
+
+    /**
+     * Test that saving an empty set with the `append` strategy on create works.
+     *
+     * @dataProvider emptyProvider
+     * @return void
+     */
+    public function testSaveAssociatedAppendEmptySetCreateSuccess($value)
+    {
+        /* @var $table \Cake\ORM\Table|\PHPUnit_Framework_MockObject_MockObject */
+        $table = $this
+            ->getMockBuilder('Cake\ORM\Table')
+            ->setMethods(['table'])
+            ->getMock();
+        $table->setSchema([]);
+
+        /* @var $assoc \Cake\ORM\Association\HasMany|\PHPUnit_Framework_MockObject_MockObject */
+        $assoc = $this
+            ->getMockBuilder('\Cake\ORM\Association\HasMany')
+            ->setMethods(['_unlinkAssociated', '_saveTarget'])
+            ->setConstructorArgs(['comments', ['sourceTable' => $table]])
+            ->getMock();
+        $assoc->setSaveStrategy(HasMany::SAVE_APPEND);
+
+        $assoc
+            ->expects($this->never())
+            ->method('_unlinkAssociated');
+        $assoc
+            ->expects($this->never())
+            ->method('_saveTarget');
```

Comment:
Instead of mocks integrating with the db and checking the db state afterwards has been resulting in easier to read and easier to maintain tests lately.